### ISSUE TITLE
Ambiguity between type ascription and absolute paths

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#64242]
I am not sure, but I think the following program should be valid:

```rust
type X = ();

fn main() {
    let ():::X = ();
}
```

It currently fails with

```
error: expected identifier, found `:`
 --> /home/matklad/tmp/main.rs:4:13
  |
4 |     let ():::X = ();
  |             ^ expected identifier
```

because we don't try to decompose `::` token. 

I am not sure what should be the behavior of 

```rust
type X = ();

fn main() {
    let Y:::X = ();
}
```

but I lean towards allowing it as well, looks like it requires only constant look-ahead to handle.

Anyway, `rg ':::' --type rust` shows that we don't have a test here.

@petrochenkov what are your thoughts on this?

**Comment by matklad** [comment]
сс @rust-lang/wg-grammar

**Comment by petrochenkov** [comment]
>what are your thoughts on this?

I was aware of this, but never planned making a fix due to low priority.
It's quite possible that some other multi-character punctuation is not broken into parts when it should.
As you could [notice](https://github.com/rust-lang/rust/issues/47856) things like this are fixed on "on demand" basis.
Hopefully, issues like this will disappear as a class after migrating to the single-character punctuation model.

**Comment by Centril** [comment]
> I am not sure, but I think the following program should be valid:

According to libsyntax (`fn parse_path`), the grammar is:
```
    /// `path = [::] segment+`
    /// `segment = ident | ident[::]<args> | ident[::](args) [-> type]`
```

A path cannot end with `::` and so an identifier is correctly expected but not found.

> I am not sure what should be the behavior of

Unless I'm missing something fundamental here, the same issue as above applies and this should remain an error.

**Comment by matklad** [comment]
@Centril I think you are trying to read this as `:: :`, while I read it as `: ::` I believe `(): ::X` and `Y: ::X` are unambiguous parses, it's just that the current implementation is to eager to eat `::`. 

I've checked with syn, and it parses the first example but not the second one.

**Comment by matklad** [comment]
> Hopefully, issues like this will disappear as a class after migrating to the single-character punctuation model.

Yeah, that's exactly how I found this. My fear here though is that a genuine ambiguity might lurk in there...

**Comment by Centril** [comment]
> @Centril I think you are trying to read this as `:: :`, while I read it as `: ::` I believe `(): ::X` and `Y: ::X` are unambiguous parses, it's just that the current implementation is to eager to eat `::`.

Oh I see. What's the ambiguity referenced in the title then?

**Comment by matklad** [comment]
@Centril that was a non-formal use of the word, in a sense that rustc's parser can confuse the two. I am almost sure that there's no ambiguity on the grammar level here with tokens-with-jointness model, but I want to double check

**Comment by Centril** [comment]
@matklad I see. For now, I think it would be good to add a test documenting the current behavior to prevent accidents.

**Comment by GrigorenkoPV** [comment]
Type ascriptions are no longer a thing. Thanks to #130673, the parser now recovers from `:::` to just `::`. Should this issue be closed?

**Comment by fmease** [comment]
> I think you are trying to read this as `:: :`, while I read it as `: ::` I believe `(): ::X` and `Y: ::X` are unambiguous parses, it's just that the current implementation is to eager to eat `::`. 

Independently of whether the parser is based on multi-character punctuation or single-character punctuation, isn't the cross-language convention to tokenize or parse, respectively, with "maximum munch" / "longest match" (cc [Wikipedia](https://en.wikipedia.org/wiki/Maximal_munch)). So `:::` should ultimately always be interpreted as `::` followed by `:`, not as `:` followed by `::` regardless of implementation?
